### PR TITLE
Move tests out of the packaged directory

### PR DESCRIPTION
**Pull request by dsanders11**
Tests inside the directory that gets packaged and put on PyPI is a personal pet peeve of mine, so I've refactored the tests so that they're outside of the packaged directory. Should be no noticeable change if I've done my job right. Git didn't recognize the move from `newsletter/tests/__init__.py` to `tests/__init__.py` probably because that filename moved to `test_project/test_project/__init__.py` so poor Git got all confused, but it's just a simple move.

This drops some of the custom code in `runtests.py` in favor of executing Django's test command directly. This means we get all of the options instead of just the ones that were being defined manually.

There's two ways to run the tests, just as before:

````
# Run all tests with no options
python setup.py tests

# Run the tests with any option available from manage.py test
python runtests.py --parallel 5

# Run individual test files
python runtests.py tests.test_mailing
````

@dokterbob, I'll merge this to master if you're cool with it.

### CODE REVIEW

**Comment by dokterbob** [comment]
Hey @dsanders11,

Thanks - as always - for the contrib. However, wait a second before merging. As usual, I have some comments. ;)

tl;dr: does this solve a problem and does the proposed solution actually solve it?

It seems that including tests does slightly increase the size of the app at the advantage of being able to test the app from within projects with different settings. For example, let's suppose there is a weird bug that occurs only with one particular kind of database. I have personally experienced similar issues and being able to test the app from within my project really does make a difference. Problems could be surfaced by simply running `./manage.py newsletter`, even though 'nowadays' the idea is to only run project-related (as opposed to reusable apps') tests. This would be a feature lost at a minimal size gain, or are there other reasons I have not been aware of?

Then about moving the tests as a way of not including them. It seems to me that simply `exclude`'ing them from within `setup.py`'s `find_packages()` would suffice. Also, given [different](http://as.ynchrono.us/2007/12/filesystem-structure-of-python-project_21.html) [opinions](http://docs.python-guide.org/en/latest/writing/structure/#sample-repository) and a default of having tests [in the app](https://docs.djangoproject.com/en/1.9/intro/reusable-apps/#your-project-and-your-reusable-app) I see no conclusive reason for moving them to the top level.

Lastly, about using Django's test runner directly: AFAIK this has only recently become possible, at last! I can only applaud such efforts and I would LOVE to see a patch that does that (and then just). 

Looking forward to hear your opinion on this. Hope you don't mind me being opiniated on these matters.

Mathijs

**Comment by dokterbob** [comment]
On an additional note: the idea seems to be to include tests in source packages but to include them from binaries, as per https://github.com/aio-libs/aiopg/issues/85. Am I understanding this right?

**Comment by dsanders11** [comment]
Having the tests outside of the source directory is a common practice among Python libraries: [django-extensions](https://github.com/django-extensions/django-extensions), [sorl-thumbnail](https://github.com/mariocesar/sorl-thumbnail), [requests](https://github.com/kennethreitz/requests), [requests-oauthlib](https://github.com/requests/requests-oauthlib), etc. There are a couple advantages, but there's the meta advantage of looking and feeling similar to other known projects, it makes it easier for others to jump in and contribute.

Off the top of my head here's a list of reasons I think this approach is superior.

#### Unwanted Bloat in Disk Space

For my overall project a quick rough estimate shows that between the `*.py` and `*.pyc`, test code from various Python projects is taking up about 9% of the overall disk space used by installed Python packages source code (which is a lot, since Django alone takes up 34%). While `django-newsletter` itself doesn't contribute very much to this overall, it's a "tragedy of the commons" problem where when every library does it it adds up to a significant total.

#### Keeps the Codebase and Git Log Clean

Putting test code in with the regular source code muddies the codebase. Extra resources and code are put into it, raising disk size as mentioned above. It can easily creep and become hard to detangle the tests from the source code. For example, `django-newsletter` has a `fixtures` folder but the only thing in it is for tests. It adds unnecessary folders and files to the source. While `find_packages` can be used to exclude packages as you mentioned, it can't be used to exclude data files like in `fixtures`. It also requires extra maintenance to ensure it's catching any new test files that may get made in a different path.

Another disadvantage of relying on `find_packages` exclude is that downloading the ZIP version of the project from the 'Releases' page will give you a different directory layout than installing from PyPI, which can be confusing.

In addition to keeping the file directories cleaner, it also creates a cleaner git log. Refactoring tests with lots of commits will not make the `git log` for the source code directory cluttered. It's easier to see how many commits changed the actual source code, rather than just commits that were improving test coverage or something.

#### Searching the Code

This is one of my bigger pet peeves about including the test code in the source code. Often times when tracking down bugs (and there are lots and lots of them in open source software) I need to track certain sections of code, so using `grep` on the codebase is very useful. However, if the test code is included with the source code, it generates a lot of extra matches that aren't helpful and that I need to manually exclude from my searching.

This also applies just to doing development on a single project like `django-newsletter`. It's often useful to search the source code, and being able to cleanly exclude all of the test code while doing that is very nice.

#### Mix and Match Tests

Having the tests be in a separate directory makes it quite easy to mix and match tests with the source code. For example, someone has a branch they've been working on for a bit which changes a good amount of source code. Since they've started the branch, more test coverage was added to master and the test code was refactored.

Say you or them want to run the latest tests from master against their branch. They can merge master into their branch and take the time to fix the merge conflicts (which may be wasted time if the branch isn't completed yet and will change). Or, with separate source code and tests directories, the following in a clean git clone would allow the tests from master to be run against the branch easily:

```
# Check out the branch version of just the source code directory
git checkout dsanders11/some-branch newsletter

# Run the tests, using tests from master and source code from branch
./runtests.py
```

The reverse can also be done, running tests from a branch against the master source code. This can be done with the current layout but it's harder since the tests are inside of the source code, and is more error prone.

#### Running Tests in Project

This would still be possible, it's just trickier to do. Since moving the tests means they use full imports (e.g. `from newsletter.models import Newsletter`) instead of relative imports, the tests can easily be run against an installed version of the app. You would need to git clone the `django-newsletter` repository, and then change one or two lines in `runtests.py`. We could try to make this easier with a special version of `runtests.py` or a flag.

IMO, running tests in a full project is bad practice. In theory it should rarely be necessary if the app has good test coverage. I also don't see it working well much of the time - I don't use several of the `django-newsletter` URL patterns (submission archive, newsletter list, etc) so the tests would fail with lots of false negatives if I tried to run the tests in my project. As such it's not a great debugging tool for anyone who doesn't know what they're doing unless the tests are made to be very resilient to missing URL patterns and other potential changes a project may have.

**Comment by claudep** [comment]
> IMO, running tests in a full project is bad practice. In theory it should rarely be necessary if the app has good test coverage. I also don't see it working well much of the time - I don't use several of the django-newsletter URL patterns (submission archive, newsletter list, etc) so the tests would fail with lots of false negatives if I tried to run the tests in my project. As such it's not a great debugging tool for anyone who doesn't know what they're doing unless the tests are made to be very resilient to missing URL patterns and other potential changes a project may have.

Agreed. We had that same experience in the Django project, where contrib apps used to be tested with user projects, and we realized at some point it was impossible to properly test things independently from the client specific context.

**Comment by dokterbob** [comment]
Ok, I'm convinced. The 9% broke the deal. 

I was particularly amused by your "tragedy of the commons" argument, just [reading up on that stuff a bit](http://wtf.tw/ref/ostrom_1990.pdf), recently.

**Comment by dokterbob** [comment]
@claudep @dsanders11 Thanks for thinking along guys. It's nice to not have to be stuck in my own preconceptions. :)
